rust/psibase_macros/psibase-macros-lib/src/service_macro/actions.rs

```diff
@@ -190,3 +190,121 @@ pub fn process_action_schema(
         actions.insert(#name_str.to_string(), #psibase_mod::fracpack::FunctionType{ params: builder.insert::<#action_mod::#name>(), result: #ret });
     }
 }
+
+pub struct PreAction {
+    pub exists: bool,
+    pub fn_name: Option<Ident>,
+    pub exclude: PathList,
+}
+
+impl Default for PreAction {
+    fn default() -> Self {
+        PreAction {
+            exists: false,
+            fn_name: None,
+            exclude: PathList::from(vec![]),
+        }
+    }
+}
+
+fn is_pre_action_attr(attr: &Attribute) -> bool {
+    if let AttrStyle::Outer = attr.style {
+        if attr.meta.path().is_ident("pre_action") {
+            return true;
+        }
+    }
+    false
+}
+
+#[derive(Debug, FromMeta)]
+struct PreActionOptions {
+    exclude: PathList,
+}
+
+pub fn check_for_pre_action(pre_action_info: &mut PreAction, items: &mut Vec<Item>) {
+    // println!("check_for_pre_action().top");
+    for (_item_index, item) in items.iter_mut().enumerate() {
+        if let Item::Fn(f) = item {
+            if f.attrs.iter().any(is_pre_action_attr) {
```

Comment:
`any` and `find` are doing redundant work. Try
```
if let Some(pre_action_attr) = find(...)
```
Or use `position` since you'll also need that later.